Add support for the yash shell

awkless opened the pull request:
## Description

I added support for the [yash shell][yash-website], which is my current working shell. Originally, the `polyglot.sh` script stated that my shell was not supported. Now the prompt comes out as desired, however I could not get yash to display the current Vi-mode, because it does not support that. I tested my modifications to ensure that it does not break existing functionality for other shells, and so far everything works.

[yash-website]: https://yash.osdn.jp/

awkless committed:
feat: Add support for the yash shell

I wanted to use the `polyglot.sh` script for my current working shell
yash, but the script stated that my shell was not supported. So i
decided to add support for it.

I did my best to adhere to the coding style, and prompt style of the
script. The only thing lacking is the ability to show what Vi mode you
are in since yash does not support that (I looked through its manual
and there is no mention of such a feature at all).

Signed-off-by: Jason Pena <jasonpena@awkless.com>

agkozak commented:
This is an excellent idea, and what you have works.

I think I can see a way to accomplish the same thing with even fewer lines of code, though. I'll probably have time on Monday to work on this. Basically, there's a way to generate the strings for `pdksh` and `yash` using the same code. I'll show you what I mean in a couple of days.

Thank you so much for contributing to this project.

awkless committed:
refactor: Remove redundant code from yash prompt

Moved yash prompt implementation into the same if-else tree used
by pdksh, dash, busybox ash, zsh in sh emulation mode. This
removed redundant code for both user and superuser prompts for yash.

Signed-off-by: Jason Pena <jasonpena@awkless.com>

awkless commented:
> I think I can see a way to accomplish the same thing with even fewer lines of code, though.

I saw what you meant and refactored my patch to remove as much of the redundancies of the original implementation.

> Basically, there's a way to generate the strings for pdksh and yash using the same code.

I see what you mean. However, I ran into a wall with Yash not having the `print` command. Although, upon looking at the code you are using for the superuser section of pdksh I have a rough idea on how to solve this in order to achieve what you are talking about.

Overall, I think I can refactor my patch to what you are envisioning over the weekend. I just need to do some experimentation.

agkozak commented:
> However, I ran into a wall with Yash not having the `print` command.

I *think* the solution is to use virtually the same code, only with `echo`. I'll have more time on Monday to spend on this issue.

agkozak commented:
Ah, looking more closely, I see that `yash` uses `bash`-style color escape sequences (i.e `\[...\]`), so what you're doing now is quite necessary. Good. All right, I'll continue to play with your feature, but I expect that I should be able to merge your changes to `master` soon. Thanks again. I'll check in with you if I have any questions.

awkless commented:
After some experimentation I have come to the following conclusions.

Yash's echo built-in is unreliable for interpreting escape sequences. You see, yash has a unique implementation of echo whereby the user can select the style of echoing through the [ECHO_STYLE][yash-echo-style] environment variable. Certain styles disallow the echo command to interpret escape sequences properly unless you explicitly use the `-e` option. However, the `RAW` style mode for echo not only disables escape sequences, but disables all options making `-e` useless. Even worse, the default configuration that yash uses when the user does not have a `.yashrc` file sets `ECHO_STYLE` to `RAW` anyway. So if the user decides they want to use yash's defaults  via `. --auto-load --no-alias initialization/common` in their `.yashrc`, along with `polyglot.sh`, then they will be left with a broken prompt.

However, even if we somehow find a way to make echo work reliably regardless of what the user sets `ECHO_STYLE` too, the pdksh strings still cause a number of issues for yash. Firstly, yash's auto completion feature gets messed up whenever you hit tab and have multiple options to choose from, e.g.,

```
Before using pdksh strings for PS1 after hitting tab
----------------------------------------------------
user@host ~ $ vim test1.c
[test1.c]  test2.c  test3.c

After using pdksh strings for PS1 after hitting tab
---------------------------------------------------
user ~ $ vim                                .c
[test1.c]  test2.c  test3.c
```
In the second example above, a strange gap is placed after whatever command you are trying to auto complete. If you try to use backspace in the hopes of fixing this oddity, then the command will disappear leaving the gap and cursor. Afterwards, you can still type, but nothing will be displayed until you hit enter.

Lastly, the pdksh strings mess up yash's [line editing prediction][yash-le-predict] feature, which is a built-in equivalent to ZSH's `zsh-autosuggest` plugin. So if you have `set --le-predict` in your `.yashrc`, then yash will begin to predict what commands you are currently typing. When using the pdksh strings for PS1, yash displays nothing for its prediction as if you never enabled line editing prediction in the first place.

All in all, I do not think we can reliably combine pdksh and yash together unless we heavily modify the pdksh strings to compensate for yash itself. 

[yash-echo-style]: https://yash.osdn.jp/doc/_echo.html
[yash-le-predict]: https://yash.osdn.jp/doc/lineedit.html#prediction

agkozak commented:
Here's what I've done over on the [`yash` branch](https://github.com/agkozak/polyglot/tree/yash):

* Accepted your two commits;
* **[Made it so that, when color is supported, a `yash` superuser will have their username@hostname appear in reverse video, as with Zsh, `bash`, `ksh93`, `mksh`, and `pdksh`](https://github.com/agkozak/polyglot/commit/9a96ddea24f3bf53dabde438cdc401012a1d329b#L725-729);**
* Removed my previous attempts to have superuser username@hostname appear in reverse for a superuser with a monochrome terminal, as experimenting today showed me that it is not working reliably enough
* Revised the `BusyBox` detection routine, as it was not working reliably on BSD

Do you think that should be sufficient?

awkless commented:
Yes, your changes work very well. No issues to report on my end.
Thanks!

awkless commented:
I noticed that you haven't merged the pull request. Are you moving all development to the new `yash` branch? If so, what else do you need implemented?

agkozak commented:
I'll merge it to `master` tomorrow and give you a heads-up when I do. Thanks again so much for the new feature and for making me aware of `yash`.

agkozak commented:
@awkless The new feature has been merged to the `master` branch and is part of [release v1.12.0](https://github.com/agkozak/polyglot/releases/tag/v1.12.0). You are listed as a contributor. Thanks again!

agkozak commented:
(I redid the animated demo so that it would have `yash` in it.)
